src/helpers/audio/audio.ts

```diff
@@ -0,0 +1,211 @@
+import { IAudioInformation, MESSAGE_PROCESSOR_READY, SpectrogramOptions, State } from "./state";
+import { IAudioMetadata, parseBlob } from "music-metadata-browser";
+import bufferBuilderProcessorPath from "./buffer-builder-processor.ts?worker&url";
+// import workerPath from "./worker.ts?worker&url";
+import WorkerConstructor from "./worker.ts?worker&inline";
+import { Size } from "../../models/rendering";
+
+interface IFetchedAudio<T = Response | ArrayBuffer> {
+  response: T;
+  metadata: IAudioMetadata;
+}
+
+export class AudioHelper {
+  public constructor() {
+    // const spectrogramWorker = AudioHelper.worker || new Worker(new URL(workerPath, import.meta.url));
+    this.spectrogramWorker = new WorkerConstructor();
+  }
+
+  private readonly spectrogramWorker: Worker | undefined;
+  private processorNode: AudioWorkletNode | undefined;
+  private state: State | undefined;
+  private sampleBuffer: SharedArrayBuffer | undefined;
+  private offscreenCanvas: OffscreenCanvas | undefined;
+  private cachedFile!: IFetchedAudio<Response>;
+  private segmentSize = 41100;
+
+  public get canvasTransferred(): boolean {
+    return this.offscreenCanvas !== undefined;
+  }
+
+  public async connect(src: string, canvas: HTMLCanvasElement, options: SpectrogramOptions): Promise<IAudioMetadata> {
+    const { response, metadata } = await this.fetchAudio(src);
+
+    const context = await this.createAudioContext(metadata, response);
+
+    // the number of samples after which to trigger a render of the spectrogram
+    // the balance of this number is a performance tradeoff
+    // - too many samples and we'll use more memory and render in larger clunky chunks
+    // - too few samples and we'll be rendering too often and hit performance bottlenecks
+    //   e.g. with canvas painting, wasm interop, signalling primitives, etc.
+    //
+    // about one seconds-worth of samples
+    this.state = State.createState(this.segmentSize);
+    this.sampleBuffer = new SharedArrayBuffer(Float32Array.BYTES_PER_ELEMENT * this.segmentSize);
+
+    if (!this.canvasTransferred) {
+      this.offscreenCanvas = canvas.transferControlToOffscreen();
+    }
+
+    this.setupWorker(metadata, options);
+    this.setupProcessor(context);
+
+    console.time("rendering");
```

Comment:
wrong spot?

Reply:
Moved to the `setupProcessor` function above `startRendering()`